### PR TITLE
Handle cycles when propagating markers

### LINKED ISSUE #4584: pip compile --universal does not retain environment markers for dependencies
**Steps to reproduce:**
`echo poetry | uv pip compile --universal -`
adding `--no-strip-markers` yields the same result.

**Outcome:**
```
...
xattr==1.1.0
    # via poetry
```
xattr is not available for windows, so the generated requirements will fail to install on windows.

**Expected outcome:**
```
...
xattr==1.1.0; sys_platform == "darwin"
    # via poetry
```
From poetry's sdist: `Requires-Dist: xattr (>=1.0.0,<2.0.0) ; sys_platform == "darwin"`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

It turns out that `Topo` only works on graphs without cycles. If a graph has a cycle, it seems to bail early. So we were losing markers for trees that contain cycles (like Poetry, which depends on `poetry-plugin-export`, which depends on Poetry).

Now, we remove cycles beforehand and re-add those edges afterwards.

It's a bit hard for me to reason about the implications of this. The way that marker propagation works is that we do visit the nodes in-order and propagate the markers from any incoming to any outgoing edges. We only do this at a single depth (rather than recursively) because we visit the nodes in-order anyway. But if you have a cycle... then in theory you might need to propagate the markers recursively? Or maybe not?

As an example:

`A -> B -> C -> D -> B`

If `A -> B` has `sys_platform == 'darwin'`, and then `D -> B` has `python_version >= '3.7`... then we don't need to propagate `python_version >= '3.7'` back to `B` or any of its dependencies, because the condition would be `(sys_platform == 'darwin' or python_version >= '3.7) or sys_platform == 'darwin'`, which is equivalent to `sys_platform == 'darwin'`.

Closes #4584.

### CODE REVIEW

**Comment by charliermarsh** [comment]
The trio resolution changed; need to check if it was incorrect before or is incorrect now.

**Comment by charliermarsh** [comment]
Ok, this is a clear improvement, but I'm going to look into doing something more rigorous.

**Review by zanieb — APPROVED** [review]

**Comment by zanieb** [comment]
This looks like a pretty reasonable start.

**Review by BurntSushi — COMMENTED** [review]
LGTM!

I can't immediately see where your approach would fail. Maybe something to do with optional dependencies? But I can't think of a way that can go wrong.

**Comment by charliermarsh** [comment]
I really wish there was a way to do this without modifying the graph structure... I've thought a lot about it and I can't find a way to do it with a traditional BFS or DFS.

File: crates/uv-resolver/src/resolution/display.rs [hunk 1]
```diff
@@ -348,6 +349,28 @@ fn to_requirements_txt_graph(graph: &ResolutionPetGraph) -> IntermediatePetGraph
...
+    // TODO(charlie): The above reasoning could be incorrect. Consider using a graph algorithm that
+    // can handle weight propagation with cycles.
+    let edges = {
+        let fas = greedy_feedback_arc_set(&graph)
+            .map(|edge| edge.id())
+            .collect::<Vec<_>>();
+        let mut edges = Vec::with_capacity(fas.len());
+        for edge_id in fas {
+            edges.push(graph.edge_endpoints(edge_id).unwrap());
+            graph.remove_edge(edge_id);
```
Comment (bluss) [inline crates/uv-resolver/src/resolution/display.rs:369]: Hm is this maybe not a a StableGraph so not safe to remove edge ids like this probably, unless you sort the edge ids?

This is a drive by comment, but petgraph caught my attention (even though it doesn't do that much anymore, sorry about that :slightly_frowning_face: )

About your general questions in this PR - the back of my mind says reverse postorder to do something similar to Topo but accomodating cycles to some extent.

An algorithm like scc https://docs.rs/petgraph/latest/petgraph/algo/fn.kosaraju_scc.html  finds both the sccs (they are cyclic) as well as a topological order between each component too.

Comment (charliermarsh) [reply crates/uv-resolver/src/resolution/display.rs:369]: Oh that’s a very smart comment — the edge IDs could change, right!

I would love help with this problem BTW…

Comment (bluss) [reply crates/uv-resolver/src/resolution/display.rs:369]: Well, I am acutely aware of petgraph's shortcomings but I'm happy that it's still useful. You managed to snipe me, so I've made a try :slightly_smiling_face:

**Pull request merged**